Use dedicated users for monasca-log-agent

kornicameister opened the pull request:
Following commit provides registering custom
project and dedicated user registered inside
this project

Commit:
Use deticate users for monasca-log-agent

Following commit provides registering custom
project and dedicated user registered inside
this project

Commit:
Added some debug details

Commit:
Added including monasca:common

Commit:
Include monasca:common elsewhere

Commit:
Removed return clause'

Commit:
Removed action:nothing

Commit:
Move log outside block and added some variables

Commit:
Added newline

Commit:
Removed block

Commit:
Added new configuration variables into conf template

Commit:
TNT

Commit:
Use group instead of 2xowner

Commit:
Removed extra " from log-agent.conf.erb

Commit:
Changed way to access values

Commit:
Added user&group into databad

jgrassler reviewed (changes requested):
A couple of things need changing (cannot comment inline, since these lines are not part of the diff):

1) https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/11/files#diff-0994b56f3bf9683da6cfa637315f68bdR27 must move to `default.rb`. Ruby is call-by-reference, so this operates on a reference to `node`. And changing `node` from a chef recipe appears to clobber it.

2) Can you move https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/11/files#diff-0994b56f3bf9683da6cfa637315f68bdL28 to `common.rb`? Since this user is a shared resource used by all agents that would be more appropriate.

Apart from these two this pull request looks good to me.

kornicameister commented:
@jgrassler I cannot say from github what exactly you want to specify with those diffs, sry :-1: 
Is there any way you could post that ? Like with maybe from line to line ?
In the meantime I will try to understand what exactly you have in mind to be moved to common.rb

jgrassler commented:
Sorry, these two links didn't work out that well indeed. I'll just link to crowbar-openstack:monasca to make my point:

1) https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/blob/2e331a693173b5d5e6609f30a27f224824315a3d/chef/cookbooks/monasca/recipes/log_agent.rb#L26 needs to move to `default.rb` (do not change the contents of `node` in chef recipes, for some reason that causes all keys to have empty values).

2) Please move all `keystone_register` resources (they start here: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/blob/2e331a693173b5d5e6609f30a27f224824315a3d/chef/cookbooks/monasca/recipes/log_agent.rb#L26 ) to `common.rb`, since the tenants/roles/users should already exist before even a single log agent is rolled out. Also, this improves performance since otherwise every log agent deployment would query Keystone multiple times to check whether the tenants/roles/users exist already.

kornicameister commented:
Regarding 1 -> I think I will just leave it empty for now, kind of pointless to put anything there until we know where and how to get that value
Regarding 2 > Ok, that's a nice point to implement thx

kornicameister commented:
Actually, regarding comment 2, am not exactly follow, because pretty much everywhere it is done like already presented in log-agent

jgrassler commented:
Regarding (1): yes, that sounds good for now.

Regarding (2): what do you mean by "everywhere"? All other barclamps? Because there the situation is usually different. There we've got one tenant/role/service user, that is used by one OpenStack service (such as Heat). Since there's only one of either it doesn't really matter where we create it. With the log-agent we've got one tenant/role/user that is used by both the log agents and by the receiving side. So it makes sense to create these Keystone resources in `common.rb` which is included by both the log_agent and the server role.

kornicameister commented:
2) Yeah, that's sound reasonable, thanks for pointing my into that explanation.

Commit:
Moved registering log-agent

Since agents (metric&logs) will be configured
with single credentials per each node it made sense
to put creating them inside the common part.

Additionally added randomizing the agents
password

Commit:
Fixing hound error

Commit:
Fixing hound error

Commit:
Fixed typo

matrixik commented on chef/cookbooks/monasca/recipes/common.rb:
```diff
@@ -0,0 +1,118 @@
+}
+
+log "monasca:common" do
+  message "monasca:common packagess are #{monasca_pkgs}"
```
packages

kornicameister replied:
Done

matrixik commented on chef/cookbooks/monasca/recipes/common.rb:
```diff
@@ -0,0 +1,118 @@
+end
+
+unless agents_settings.empty?
+  agents_settings.each do |hulk|
```
Is `hulk` some kind of acronym?

kornicameister replied:
Big green guy whose power is so great that he made himself to crowbar-openstack :anger:

kornicameister replied:
To be honest, changed to ```as``` short for ```agent_settings```

matrixik commented on chef/cookbooks/monasca/recipes/common.rb:
```diff
@@ -0,0 +1,118 @@
+
+log "monasca:common" do
+  message "monasca:common packagess are #{monasca_pkgs}"
+  level :debug
```
Do we want `:debug` here? (similar in lines 36 and 40) 
Maybe it should be moved to separate variable in `default.rb` file.

kornicameister replied:
Removed :level entirely, will be just plain call to log method

matrixik reviewed (dismissed):
For my unskilled eye besides few small comments looks fine. I still did not tested it.

Commit:
Addressed matrixik comments

Commit:
Fixing undefined method in monasca_service.rb

Commit:
Use empty strings instead of null values

jgrassler commented on chef/cookbooks/monasca/recipes/common.rb:
```diff
@@ -0,0 +1,110 @@
+unless agents_settings.empty?
+  agents_settings.each do |as|
+
+    keystone_register "monasca:common #{as["service_user"]} in #{as["service_tenant"]} user" do
```
There are two problems with this resource:

1) `as["service_user"]` and `as["service_tenant"]` are empty for some reason. You can see this in the [gist](https://gist.githubusercontent.com/kornicameister/b35ad7dac3192b5e785186cf08e70e19/raw/a651d3945d66f1d6eb5e322d999c67916505426f/stack_trace2.out) you sent me: compare the code in `common.rb` and what actually gets passed to the `keystone_register` (in the bottom portion of the gist). The `tenant_name` and `role_name` parameters do not show up. When checking the proposal they were empty strings. At this point, my best guess is that setting a default in `attributes/default.rb` only has an effect if the attribute in question is unset in the data bag (rather than being an empty string). There'd be two solutions to this: provide a sensible value in the data bag or omit these two from `template-monasca.json` and set them to `required: false` in `template-monasca.schema`. In this case the values from `attributes/default.rb` should end up being used.

2) I manually set the missing `service_user` and `service_tenant` attributes in the data bag. At this point I got another failure because the service tenant didn't exist. Inserting the following resource above this line should fix this:

```
keystone_register "monasca:common create agent tenant #{as["service_tenant"]} for monasca" do
  protocol keystone_settings["protocol"]
  insecure keystone_settings["insecure"]
  host keystone_settings["internal_url_host"]
  port keystone_settings["admin_port"]
  auth register_auth_hash
  tenant_name as["service_tenant"]
  action :add_tenant
end

```

Commit:
trying to make it work

Commit:
trying to make it work 2

Commit:
trying to make it work 3

Commit:
trying to make it work 4

Commit:
trying to make it work 5

Commit:
trying to make it work 6

Commit:
polished

Commit:
Pleasing Hound

jgrassler commented:
Uhm. Why did you merge this already? I wasn't finished reviewing, yet. It looked mostly good, but there was one crucial thing still missing, see the second point of https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/11#pullrequestreview-20376770 (you'll need to click on the "Show outdated" button).

Nothing creates the `cmm` tenant as far as I can tell (at least that was the case yesterday). I'll test this against a fresh cloud (building right now) to verify and comment again once I've run the barclamp on my cloud.

samirjorina commented:
Well, we tested this and all went fine, so we decided to merge that... In case you have still some comments, please let us know, we will take them into account before creating the PR to crowbar/crowbar-openstack.

As for the `cmm` tenant - IMHO it should be created by the Ansible installer...

jgrassler commented:
Every Keystone resource apart from the tenant is created by `common.rb` already. If we create the tenant in `common.rb` as well, the Ansible installer won't have to talk to Keystone at all and we save a bunch of extra parameters (the Keystone credentials in `register_auth_hash`) we'd have to pass into the Ansible installer otherwise.

jgrassler commented:
Alright, never mind. I applied the barclamp now and https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/11/files#diff-64945423c76c157caa7e02438b49229cR45 creates the tenant just fine, so looks like whatever broke that resource yesterday got fixed in this morning's series of commits. I think 4d25b79b73db6 took care of it by fixing the value of `monasca_project`.
